numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2694,171 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
     return c;
 }
 
+
+static void
+array_dlpack_deleter(DLManagedTensor *self)
+{
+    PyArrayObject *array = (PyArrayObject *)self->manager_ctx;
+    // This will also free the strides as it's one allocation.
+    PyMem_Free(self->dl_tensor.shape);
+    PyMem_Free(self);
+    Py_XDECREF(array);
+}
+
+/* This is exactly as mandated by dlpack */
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
+    if (managed == NULL) {
+        return;
+    }
+    managed->deleter(managed);
```

Comment:
```suggestion
    if (managed->deleter) { // DLPack header says this can be NULL... 
        managed->deleter(managed);
    }
```

Reply:
Yes, but this deleter is only needed for the capsules NumPy produces, and NumPy never produces capsules with a `NULL` deleter.

Reply:
oh yes we export it.